keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+
+The DRA driver kubelet plugin receives that information when the pod is about to
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation types.
+
+### Publishing node resources
+
+A semantic model defines an attributes type. This is a Go type that gets
+embedded inside the resource.k8s.io API, not a separate API kind. The kubelet
+queries the registered DRA driver kubelet plugins for resource information
+through a new gRPC call. A streaming gRPC API allows drivers to update the
+capacity information.
+
+For example, a simple counter could be described in a `counter` Go package as:
+```
+type Attributes struct {
+    // Some meta information just for logging and debugging. Not used for counting.
+	metav1.ObjectMeta
+
+	Count int64
+}
+```
+
+The kubelet then publishes this information in a new NodeResources object on
+behalf of the driver. This implies that kubelet, DRA driver, and apiserver must
+all support the same semantic models and the same versions of them. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+### Using semantic parameters as claim parameters
+
+The semantic model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	Required int64
+}
+```
+
+The `Parameters.Required` value gets compared against `Attributes.Count`.
+
+Suppose a vendor has a CRD which allows requesting a certain number of `frobnicators`:
+
+```
+kind: ClaimParameters
+apiVersion: dra.frobincatorInc.example.com/v1alpha1
+metadata:
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
+  uid: abcd
+parameters:
+  - counter:
+      required: 100
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list
+and watch the custom claim parameter resources. This is necessary because it is
+not known in advance which CRDs will be used by DRA drivers. When deploying a
+driver, suitable role bindings for
+[`system:kube-scheduler`](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles)
+have to be added which grant read, watch and list permissions for the vendor
+CRD.
+
+The scheduler does not need to interpret the custom data in those CRDs. It
+still needs to read them for two purposes:
+- Ensure that the in-tree claim parameters match the vendor claim parameters
```

Comment:
So, if we capture the `generation` in the `generatedFrom`, something that can read the CR will be able to verify that it has not changed since the common ClaimParameters were created.

For the snapshot piece below, how is that captured? I see the resource handle in the main DRA KEP, but that is for the driver to write opaque data, not for this.

I think you are trying to capture "as-was" data - that is, the claim parameter object "as it was at the time of allocation". Because maybe someone changes it between allocation and driver actuation? That is an interesting problem. Users are annoying :)

This is a general question of whether the API is actually declarative or not. A true declarative API would adjust to user changes to the spec. I don't know if that's really what we want. I think in other areas of K8s, when using the `Class/Claim` pattern, this is also somewhat broken. At least, for Gateway API I am pretty sure classes are not declarative in the sense that the user can change them and have the change ripple through to all instances based on that class. Instead, the class is intended to be immutable.

Hmm. Hmm. I am trying to think about how we can avoid the scheduler needing read access to all these CRs. In fact read/write, because may it has to create duplicates or something. That's ugly. Stuff them away in RawExtensions inside the allocation result? I don't like that either.

I am tempted to say that the allocation result should just record the `generatedFrom` including the `generation`. If the generation has changed between allocation and actuation, the resource allocation fails (think of it as an out-of-band change - like if someone ssh'd into the box and mucked with the card configuration). If we did that, the scheduler would only need to be able to read the common claim parameter object written by the driver controller, not any CRs.

Reply:
It seems like a good simplification because users shouldn't be doing that in general.

Reply:
Maybe a webhook that reject mutation of the CR if the claim has been used? That has its own issues (permissions and otherwise).

Reply:
See also https://github.com/kubernetes/enhancements/pull/4384#discussion_r1477019579

> I think you are trying to capture "as-was" data - that is, the claim parameter object "as it was at the time of allocation". Because maybe someone changes it between allocation and driver actuation? That is an interesting problem. Users are annoying :)

The core KEP talks a bit about "resizing" in this context. It's not unthinkable that users do mutate the claim parameters and then expect that to have an effect on the running pod. The core KEP explicitly states that class and claim parameters only matter at the time of allocation. All changes after that are irrelevant and ignored.

If a DRA driver wants to support resizing, they can by watching out for changes after allocation and then comparing against the state captured during allocation. It's not covered here yet, but might become a future extension.

> Stuff them away in RawExtensions inside the allocation result? I don't like that either.

That's my current thinking.

> Maybe a webhook that reject mutation of the CR if the claim has been used? That has its own issues (permissions and otherwise).

And race conditions. Even if mutation is prevented, there's always delete+recreate to get the same effect.